Add responseBody field to InvalidResponseCodeException

chrisfillmore opened the pull request:
Issue #6853

Previously the body of an HTTP response was not included in an
InvalidResponseCodeException. This change adds a field to store
the response body, and updates DefaultHttpDataSource to populate
this value.

Commit:
Add responseBody field to InvalidResponseCodeException

Issue #6853

Previously the body of an HTTP response was not included in an
InvalidResponseCodeException. This change adds a field to store
the response body, and updates DefaultHttpDataSource to populate
this value.

chrisfillmore commented:
I haven't done any testing yet. Also, looking at DefaultHttpDataSourceTest, I see there's only one test. Please advise if you would like me to write a test to cover this new behaviour. Thanks.

julioz commented:
Please update the extension DataSource's too ([OkHttp](https://github.com/google/ExoPlayer/blob/dev-v2/extensions/okhttp/src/main/java/com/google/android/exoplayer2/ext/okhttp/OkHttpDataSource.java#L236), [Cronet](https://github.com/google/ExoPlayer/blob/dev-v2/extensions/cronet/src/main/java/com/google/android/exoplayer2/ext/cronet/CronetDataSource.java#L451)) so other callers can also benefit from the ability to inspect the error response body.

chrisfillmore commented:
@julioz is this possible for CronetDataSource? It seems response body is not exposed by `UrlResponseInfo`. Do you have a recommendation here?

I can make the change for OkHttp. Thanks.

Commit:
Update OkHttpDataSource with response body on bad HTTP status

This change supplies the response body in OkHttpDataSource to
InvalidResponseCodeException.

Issue #6853

ojw28 commented:
> @julioz is this possible for CronetDataSource? It seems response body is not exposed by UrlResponseInfo. Do you have a recommendation here?

I didn't try, but I'd guess you read it in the same way as the response body in the case of success. Try using `CronetDataSource.readInternal`, although note that you'll have to be sure to keep calling it until you've read everything.

julioz commented:
@chrisfillmore I haven't looked in detail, but I believe it is similar to how the success response is read via reading the byte stream.
Maybe @ojw28 can chime in and point the way, but at least by having the default and okhttp data sources exposing the body we're likely covering most use cases :+1:

chrisfillmore commented:
Here's what it would look like in Cronet:

```
if (responseCode < 200 || responseCode > 299) {
      ByteBuffer responseBuffer = ByteBuffer.allocateDirect((int)responseInfo.getReceivedByteCount());
      try {
        readInternal(responseBuffer);
      } catch (HttpDataSourceException e) {
        throw new OpenException(e, dataSpec, Status.INVALID);
      }

      InvalidResponseCodeException exception =
          new InvalidResponseCodeException(
              responseCode,
              responseInfo.getHttpStatusText(),
              responseInfo.getAllHeaders(),
              dataSpec,
              responseBuffer.array());
      
```
I am concerned about any side effects from `UrlRequestCallback` when calling `readInternal`.

Cronet is not (yet) one of our use cases. I'm inclined not to add this change to CronetDataSource at this time, since I don't think I can adequately certify that it works (at least, in the time I have).

@ojw28 would it be acceptable to add this to DefaultHttpDataSource and OkHttpDataSource, without CronetDataSource? For now anyway.

Commit:
Read response body in CronetDataSource for error response

Issue #6853

ojw28 commented:
> Cronet is not (yet) one of our use cases. I'm inclined not to add this change to CronetDataSource at this time, since I don't think I can adequately certify that it works (at least, in the time I have). @ojw28 would it be acceptable to add this to DefaultHttpDataSource and OkHttpDataSource, without CronetDataSource? For now anyway.

If the field is being added then it seems quite confusing to not populate it for one implementation. So I think it does need doing. I assume you've done some local testing to see that your changes in the Default and OkHttp implementations work correctly. It should easy to swap in the Cronet implementation and do the same local testing with that one as well? If you could do that and add the Cronet change to the pull request, that would be great.

If your concern is more around subtle issues that local testing might not catch, then please include the change in the pull request regardless. All pull requests go through a fairly detailed internal code review process before they're merged. So whilst you should obviously do some local testing and not send something that just doesn't work at all, you can send something in the knowledge that there's an additional safety net that has a good chance of catching any remaining issues. We can also modify the change during that review process, including (in the worst case!) omitting the Cronet part if we're not confident either :).

chrisfillmore commented:
Sorry for the delay; I will be following up with this. I just had a few priorities to deal with this week, and I'll get back to it next week.

chrisfillmore commented:
@google-oss-bot please keep the issue open, thanks :)

(I was out sick most of last week, so sorry googoobot)

ojw28 commented:
@marcbaechinger - Could we exclude Pull Requests from the Google OSS Bot? In most cases it's unwanted, and in any case the volume of PRs is low enough we can trivially do it manually.

chrisfillmore commented:
I have confirmed, through local testing, that DefaultHttpDataSource and OkHttpDataSource work as expected. Cronet testing is ongoing; not immediately working as I expect, so I will take another look tomorrow.

Also I see I have conflicts now. I will resolve those too.

chrisfillmore commented:
I'm on vacation for a week as of tomorrow. I resolved the merge conflicts but I'm still working on Cronet. I'll get back to this next week.

chrisfillmore commented on extensions/cronet/src/main/java/com/google/android/exoplayer2/ext/cronet/CronetDataSource.java:
```diff
@@ -443,16 +443,42 @@ public long open(DataSpec dataSpec) throws HttpDataSourceException {
     }
 
+    // Calculate the content length.
+    if (!isCompressed(responseInfo)) {
```
This block was moved up here so that we could know the length of the response. Besides being moved, the block is unchanged.

chrisfillmore commented on extensions/cronet/src/main/java/com/google/android/exoplayer2/ext/cronet/CronetDataSource.java:
```diff
@@ -568,6 +581,43 @@ public int read(byte[] buffer, int offset, int readLength) throws HttpDataSource
+
+  // Internal methods.
+
+  private int readIntoByteBuffer(ByteBuffer buffer) throws HttpDataSourceException {
```
This diff kind of sucks.

~None of the above methods have changed (except `read(ByteBuffer buffer)`. The change here is that there is a new private method called `readIntoByteBuffer(ByteBuffer buffer)`, which is used in two places: `read(ByteBuffer buffer)` and in `open()` (to read the error response body).~

~`readIntoByteBuffer(ByteBuffer buffer)` is the same as `read(ByteBuffer buffer)` used to be, except without the assertion that the request has been opened.~

Disregard the above, I've made an update which makes more sense, I think. See new comments.

chrisfillmore commented:
I have added logic to CronetDataSource to read the response body. Please review. Thanks!

chrisfillmore commented on extensions/cronet/src/main/java/com/google/android/exoplayer2/ext/cronet/CronetDataSource.java:
```diff
@@ -682,6 +651,50 @@ protected UrlResponseInfo getCurrentUrlResponseInfo() {
 
   // Internal methods.
 
+  private int readIntoByteArray(byte[] buffer, int offset, int readLength) throws HttpDataSourceException {
```
This new method is just the body of `read(byte[] buffer, int offset, int readLength)`, without the assertion, which remains in the public read method.

This is now called from `open()` to get the response body.

icbaker commented:
Thanks for the contribution!

> Also, looking at DefaultHttpDataSourceTest, I see there's only one test. Please advise if you would like me to write a test to cover this new behaviour. Thanks.

Adding some tests for this would be great! I think it probably makes sense to add a new test method to `DefaultHttpDataSourceTest`, the existing method in there is already doing too much imo. Something like `DefaultHttpDataSourceTest#open_invalidResponseCode()` and just assert that all the fields in the `InvalidResponseCodeException` are as you'd expect. Same for `OkHttpDataSourceTest`.

For `CronetDataSourceTest` you can probably extend `testRequestOpenValidatesStatusCode` rather than writing new method. In fact it would be great if you could take a look at this method anyway since when I patch this PR & run it the test hangs forever, blocked on this line:
https://github.com/google/ExoPlayer/blob/ab21f885d42209d0c65a70fc93dee26c96af6f84/extensions/cronet/src/main/java/com/google/android/exoplayer2/ext/cronet/CronetDataSource.java#L761

I haven't spent much time getting to the bottom of what's going on - but it's probably useful to be aware that `ConditionVariable#block(int)` doesn't timeout correctly under Robolectric (both `com.google.android.exoplayer2.util.ConditionVariable` and `android.os.ConditionVariable` because both rely on `android.os.SystemClock.elapsedRealtime()` which doesn't advance "normally" in Robolectric).

chrisfillmore commented:
Thanks for the feedback. I am not ignoring this PR, just need to complete some other work first, and I will revisit this (later this week or next week).
